tower/src/buffer/service.rs

```diff
@@ -47,11 +46,11 @@ where
     /// [`Poll::Ready`]: std::task::Poll::Ready
     /// [`call`]: crate::Service::call
     /// [`poll_ready`]: crate::Service::poll_ready
-    pub fn new(service: T, bound: usize) -> Self
+    pub fn new<T>(service: T, bound: usize) -> Self
     where
-        T: Send + 'static,
-        T::Future: Send,
-        T::Error: Send + Sync,
+        T: Service<Request, Future = F> + Send + 'static,
+        F: Send,
+        T::Error: Into<crate::BoxError> + Send + Sync,
```

Comment:
I wonder if it's more or less onerous to take a `thiserror` dependency and explicitly handle the error cases than requiring boxing. Note, for instance, that `anyhow::Error` does not impl `Into<BoxError>`; and we don't really have any requirements internally (i.e. on the `Service trait` that `T::Error` impl `std::error::Error`.

Reply:
could be worth considering but I think we should address that in a separate PR?

Reply:
opened #655 for that!